src/lib/io/sol-ipm.c

```diff
@@ -0,0 +1,323 @@
+    struct sol_ipm_receiver_entry *receiver;
+    struct mainloop_event me;
+
+    /* id == 0 messages. Those messages signal sender
+    * core that blob was consumed and may be freed */
+    if (id == SOL_IPM_ID_FREE_REMOTE) {
+        me.cb = ipm_reaper_process;
+        me.data = remote_blob;
+        sol_mainloop_event_post(&me);
+        return;
+    }
+
+    //TODO rethink this. What if we receive another message with same id
+    //before processing current one?
+    me.cb = ipm_receiver_process,
+    me.data = (void *)id; /* I could send `receiver` entry, and avoid yet
+                             another loop to find receiver. Although currently,
+                             Zephyr implementation does process events before
+                             timeouts, I'm not sure we can count on that
+                             `ipm_vectors` is not going to be changed in the
+                             meanwhile.*/
+
+    SOL_VECTOR_FOREACH_IDX (&ipm_receivers, receiver, idx) {
+        if (receiver->id == id) {
+            receiver->last_remote_blob = remote_blob;
+            int r = sol_mainloop_event_post(&me);
+            return;
+        }
+    }
+
+    /* Is it really OK/safe to print from isr? */
+    SOL_INF("Got message with id %" PRIu32 " but there's no receiver associated to it",
+        id);
+}
+
+SOL_API int
+sol_ipm_set_receiver(uint32_t id,
+    void (*receive_cb)(void *cb_data, uint32_t id, struct sol_blob *message),
+    const void *data)
+{
+    int idx;
+    struct sol_ipm_receiver_entry *receiver;
+
+    SOL_INT_CHECK(id, > ipm_max_id, -EINVAL);
+    SOL_INT_CHECK(id, == 0, -EINVAL);
+
+    SOL_VECTOR_FOREACH_IDX (&ipm_receivers, receiver, idx) {
+        if (receiver->id == id) {
+            if (receive_cb) {
+                receiver->receive_cb = receive_cb;
+                return 0;
+            } else {
+                return sol_vector_del(&ipm_receivers, idx);
+            }
+        }
+    }
+
+    /* Not changing or deleting. Must have a receive_cb*/
+    SOL_NULL_CHECK(receive_cb, -EINVAL);
+
+    receiver = sol_vector_append(&ipm_receivers);
+    SOL_NULL_CHECK_MSG(receiver, -ENOMEM, "Could not add receiver to id %" PRIu32,
+        id);
+
+    receiver->id = id;
+    receiver->receive_cb = receive_cb;
+    receiver->data = data;
+
+    return 0;
+}
+
+SOL_API int
+sol_ipm_send(uint32_t id, struct sol_blob *message,
+    void (*message_consumed_cb)(void *cb_data, uint32_t id, struct sol_blob *message),
+    const void *data)
+{
+    int r;
+    struct sol_ipm_sent_message *sent;
+
+    SOL_INT_CHECK(id, > ipm_max_id, -EINVAL);
+    SOL_INT_CHECK(id, == 0, -EINVAL);
+    SOL_NULL_CHECK(message, -EINVAL);
+
+    sent = sol_vector_append(&ipm_sent_messages);
+    SOL_NULL_CHECK(sent, -ENOMEM);
+
+    r = ipm_send(ipm_sender, 1, id, &message, sizeof(message));
+    SOL_INT_CHECK_GOTO(r, > 0, fail);
+
+    sent->id = id;
+    sent->message_consumed_cb = message_consumed_cb;
```

Comment:
indeed... and AFAIU one must "own the channel/id", right? Then this would even help. Just be sure you handle set(NULL) from the actual callback to address the "temporary owner" case: own/set -> send -> wait -> unset.